### DISCUSSION TITLE
Support for `KHR_mesh_quantization` gltf extension

### DISCUSSION THREAD

**Discussion by jameskane05** [mozilla/hubs#3622] [General Questions & Discussion] [answered]
Happy holidays! No rush for anyone to reply here! But I was looking at recent/upcoming three.js PRs and noticed support for this gltf extension - "KHR mesh quantization" -  basically reducing precision of stored values of vertices from FLOAT to 16- or 8-bit vals: https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Khronos/KHR_mesh_quantization

> "As an example, a static PBR-ready mesh typically requires POSITION (12 bytes), TEXCOORD (8 bytes), NORMAL (12 bytes) and TANGENT (16 bytes) for each vertex, for a total of 48 bytes. With this extension, it is possible to use SHORT to store position and texture coordinate data (8 and 4 bytes, respectively) and BYTE to store normal and tangent data (4 bytes each), for a total of 20 bytes per vertex with often negligible quality impact."

~60% memory savings per vertex - sounds incredible. Anybody familiar with this? Seems similar to Draco compression, but I'm thinking it would be easier to implement into hubs given it's an official gltf extension rather than a separate file format. The PR explaining mesh quantization for the gltf spec was very helpful on this point: https://github.com/KhronosGroup/glTF/pull/1673

> "The extension is specified to make implementations trivial; this is "de-facto" supported by Babylon.JS and Three.JS in that files that use 8/16-bit attributes can be deserialized by the respective loaders without further changes. I'd expect in general any compliant loader to be able to support this with minimal effort."

And indeed, the three.js changes required to support it are trivial: https://github.com/mrdoob/three.js/pull/17923/files

So after that it's just a matter of encoding correctly. Blender doesn't seem to support it yet but sounds like there's been at least one tool, gltfpack, that's implemented encoding since last year: https://www.npmjs.com/package/gltfpack

Possible we'd need to do the quantization encoding in Spoke since there's an export step there? It looks like there are some changes to the official three GLTFExporter were needed - but again, relatively minor: https://github.com/mrdoob/three.js/pull/20762/files

**Comment by robinkwilson** [comment, marked as answer]
This seems like a feature request to me! Can you make an issue/feature request (https://github.com/mozilla/hubs/issues) and link this discussion thread in it?
